### PR TITLE
kodi-waitonnetwork: enabled by default with 10 sec timeout / use NTP for Online Status

### PR DESCRIPTION
**Pull request by mglae**
There are two issues with kodi-waitonnetwork.service:

1) It is not enabled by default. Users do complain about not getting updates, failing connections to MySQL databases or IPTV services ...

2) Network Online status is not detected. Connmand-wait-online already succeed when network interface is in "Ready" state, i.e. when a ethernet card has a link status.

With this PR kodi-waitonnetwork.service is enabled by default and set to 10 seconds. This should be enough for most networks and is only a short delay if being offline. The time can still be changed manually to larger values or the service disabled.

The default delay is enabled too if updating a not already configured system.

The definition of network being online is now "when NTP time is synchronized". Wait-time-sync is extended with an optionally timeout parameter to support the waiting time. 

~~While WIP the LibreELEC-settings change is included as patch against 29bf2370.~~
Need to be committed together with LibreELEC/service.libreelec.settings#210.

### CODE REVIEW

**Comment by CvH** [comment]
The boot too fast for network up problem are likely the longest standing major problem at LE. 

If this is a good fix (i can't tell) this is very welcome !

**Comment by HiassofT** [comment]
Enabling wait-for-network by default and moving to wait-time-sync are very good ideas.

I'd keep the timeout of 10 seconds though, managed switches and wifi networks can be slow to come up (my HP 1810 needs about 5sec until link is up) and it's better to have the small hit in the rare case when kodi is used without net and a user didn't disable the wait-on-network setting than telling users they have to increase the timeout (that'd be the same as we have now, just a bit less often).

**Comment by HiassofT** [comment]
BTW: couldn't test this PR yet, will do that in the next days

**Comment by CvH** [comment]
just asking, isn't the root cause that kodi tries to update the repo and never tries it again if it fails ?

**Comment by HiassofT** [comment]
LE settings also doesn't retry. LE/kodi is pretty much broken if you have no net (and correct time set) at startup.

Making sure we have net (and time) by default is the right thing to do. People running offline can wait 10 sec or disable it in settings - but for everyone else things will just work

**Comment by mglae** [comment]
> just asking, isn't the root cause that kodi tries to update the repo and never tries it again if it fails ?

Leia is checking at start and retrying after 24 hours if repo update succeed or not.

In Matrix behavior is improved, failed updates are retried after 1 hour and repositories can define their update interval.

> I'd keep the timeout of 10 seconds though, ...

Technically correct but keep in mind that "Waiting for Network" is the first LE experience a new user does have after installation.

**Comment by HiassofT** [comment]
> Technically correct but keep in mind that "Waiting for Network" is the first LE experience a new user does have after installation.

This is why it should not be defaulted to on in the image but enabled by LE settings or first-run-wizard - after user had a chance to configure network.

didn't have time to check your code if you already do that or not

**Comment by CvH** [comment]
> In Matrix behavior is improved, failed updates are retried after 1 hour

ah okay, so users just need to wait an hour before they can use the addons, sounds like a good idea ;D tx for clarification

> I'd keep the timeout of 10 seconds

10 sec is a bit much, ~5seconds should be enough for 90+ percent of the users without wasting too much time. Never needed more then 5secs even with WLAN. Would rather start low and see what happens in the wild. We can increase if it is not enough.

**Comment by mglae** [comment]
> This is why it should not be defaulted to on in the image but enabled by LE settings or first-run-wizard - after user had a chance to configure network.

Using the wizard will break enabling the feature on updating a not configured system to LE10.

I've added a dependency to the service to check the wizard configuration file.

> Would rather start low and see what happens in the wild. We can increase if it is not enough.

It was not possible before to detect if the default value was used. I've added a flag to make the check possible in the future.

**Comment by mglae** [comment]
Description of the Leia update algorithm in [my comment](https://github.com/LibreELEC/LibreELEC.tv/pull/4802#issuecomment-751405030) was wrong, just saw the log:
```
DEBUG: CRepositoryUpdater: previous update at 2020-12-27 01:14:13, next at 2020-12-28 01:14:13
```
Commen is updated.

**Comment by HiassofT** [comment]
I did a quick test with this PR on an RPi2 connected to a Netgear GS108 switch, local NTP server on LAN announced by IP in DHCP config.

kodi-waitonnetwork.service needed 3.4 seconds, wait-time-sync.service (which doesn't seem to be tied into this PR) finished a bit later.

`systemd-analyze plot`: https://www.horus.com/~hias/tmp/libreelec/rpi2.svg
`journalctl -o short-monotonic -a` http://ix.io/2K72

So, as I already stated before, I suggest we stick to the current 10 second timeout - 4 seconds is too marginal

**Comment by mglae** [comment]
> kodi-waitonnetwork.service needed 3.4 seconds, wait-time-sync.service (which doesn't seem to be tied into this PR) finished a bit later.

wait-time-sync is using a poll frequency of 3 Hz, detection of sync can be delayed up to 333 ms.

wait-time-sync.service and wait-time-sync.target are independent and have no timeout.

> So, as I already stated before, I suggest we stick to the current 10 second timeout - 4 seconds is too marginal

I'm still afraid to many users will disable the feature by just reading the 10 seconds message.

**Comment by chewitt** [comment]
> I'm still afraid to many users will disable the feature by just reading the 10 seconds message.

Then don't show them the message?

**Comment by CvH** [comment]
I am also not sure we should default to the worst possible situation here. A middle ground should fix it for the most of the users without hurting too much. There are likely ppl that need 20sec and others that are okay with 2-3 secs. 
At my Nuc I have 3secs timeout and it boots directly into TV playback via LAN without any noticeable timeouts.

**Comment by lrusak** [comment]
What about just going back to using the default connman `connman-wait-online.service`?

**Comment by mglae** [comment]
> Then don't show them the message?

Yes, done.

> I am also not sure we should default to the worst possible [...] At my Nuc I have 3secs timeout and it boots directly

If you don't see kodi-waitonnetwork.service failing it will boot in the same time wiith a larger timeout. 

> What about just going back to using the default connman connman-wait-online.service?

Makes no difference regarding the default timeout discussion.

In best case NTP synchronization is performed within 1 sec after the network interface is configured. But on all the RTC less devices you now see an error during boot if system time is not set.

**Comment by mglae** [comment]
Rebased (maybe unneeded) and LibreELEC-settings updated against 29bf2370.

**Comment by mglae** [comment]
Removed LibreELEC-Settings patch and created LibreELEC/service.libreelec.settings#210.

Ready to go.

**Comment by CvH** [comment]
merged settings addon pr, can you add the settings addon bump here pls ?
https://github.com/LibreELEC/service.libreelec.settings

**Comment by mglae** [comment]
Thanks, done.

**Review by CvH — APPROVED** [review]

**Comment by CvH** [comment]
@mglae from slack
```
[FAILED] Failed to start Wait on network.
See 'systemctl status kodi-waitonnetwork.service' for details.
```
> this is from latest master (just updated), even though I have working ethernet (cable) connection

**Comment by mglae** [comment]
`journalctl -o short-precise|paste` please.

**Comment by jernejsk** [comment]
journalctl: http://ix.io/2Lbg
systemctl status: http://ix.io/2Lbh

**Comment by mglae** [comment]
Thanks. 

Summary of log messages:

| abs. time | rel. time | log message |
| ---: | ---: | :--- |
| :00:53.05 | -0.94 | LibreELEC connman-vpnd[2506]: eth0 {newlink} index 3 operstate 2 <DOWN> |
| :00:53.99 | 0.00 | LibreELEC systemd[1]: Starting Wait on network... |
| :00:55.93 | 1.96 |  LibreELEC kernel: IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready |
| :01:04.08 | 10.09 |  LibreELEC systemd[1]: kodi-waitonnetwork.service: Failed with result 'exit-code'. |
| :01:04.17 | 10.18 |  LibreELEC connmand[2705]: eth0 {add} route 0.0.0.0 gw 10.42.0.1 scope 0 <UNIVERSE> |
| :01:04.19 | 10.20 | LibreELEC connmand[2705]: ntp: adjust (jump): +3621918.536234 sec |

NTP adjust was missed by 0.11 sec but everything is working as configured.

---

Something different, I've noticed:
```
Nov 26 18:00:53.424153 LibreELEC systemd-udevd[2529]: eth0: Process '/usr/sbin/ethtool -s eth0 wol g' failed with exit code 75.
```
Is WOL not supported on this HW?

**Comment by jernejsk** [comment]
I'm not sure, probably not. We just started working on suspend/resume functionality.

**Comment by jernejsk** [comment]
@mglae is it possible that ntp is started a bit later on some platforms? wait on network almost always fails for me but I can ssh into LE before that timeout. I received plenty of complaints about slow startup on forum...

**Comment by mglae** [comment]
Again journalctl will tell.

Usually network setup can be slow but NTP sync succeed fast after it. In your sample within 0.02 seconds.

This PR favors slower startup over failing network based services like addon updates.

**Comment by jernejsk** [comment]
http://ix.io/2Ox5 but I don't think it's any different that one above. Anyway, people are unhappy about this delay.

**Comment by mglae** [comment]
> but I don't think it's any different that one above.

Yes, again ~9 sec between eth0 link up and configured network. The question is how to speed up the process.

We can start with enabling connman debug logging. 

Create `/storage/.config/system.d/connman.service.d/debug.conf`:
```
[Service]
Environment=CONNMAN_DEBUG=--debug
```

Please use `journalctl -o short-precise|paste` for precise time stamps.

> Anyway, people are unhappy about this delay.

This is not complete unexpected. ;-)

**Comment by jernejsk** [comment]
detailed log: http://ix.io/2OxE

This one particularly is Tritium H5 (Allwinner H5 SoC), but I experienced same thing on Allwinner H6 and H3 and on forum, one user said that same long boot is present on Amlogic (didn't say which SoC or board).

**Comment by mglae** [comment]
There is separate env variable to activate DHCP logging, please add: 
```
Environment=CONNMAN_DHCP_DEBUG=1
```

**Comment by jernejsk** [comment]
this is with DHCP debug enabled: http://ix.io/2OF0

**Comment by jernejsk** [comment]
also system-analyze output: https://imagebin.ca/v/5qyRYMWnOM9R

**Comment by mglae** [comment]
> Still doesn't help. At this point I'm starting to think that it's driver issue somehow...

Or a race in configuration. An arbitrary summary:
```
Feb 07 18:20:29.538371 LibreELEC kernel: dwmac-sun8i 1c30000.ethernet: Powering internal PHY

Feb 07 18:20:30.184905 LibreELEC systemd[1]: Starting Wait on network...

Feb 07 18:20:30.572720 LibreELEC kernel: dwmac-sun8i 1c30000.ethernet eth0: configuring for phy/mii link mode

Feb 07 18:20:30.573122 LibreELEC connmand[2411]: Adding interface eth0 [ ethernet ]

Feb 07 18:20:30.574181 LibreELEC connmand[2411]: eth0 {newlink} index 3 address 02:01:71:46:0B:0C mtu 1500
Feb 07 18:20:30.574208 LibreELEC connmand[2411]: eth0 {newlink} index 3 operstate 2 <DOWN>

Feb 07 18:20:32.623760 LibreELEC kernel: dwmac-sun8i 1c30000.ethernet eth0: Link is Up - 100Mbps/Full - flow control rx/tx
Feb 07 18:20:32.628484 LibreELEC kernel: IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready

Feb 07 18:20:32.621740 LibreELEC connmand[2411]: eth0 {add} route fe80:: gw :: scope 0 <UNIVERSE>

Feb 07 18:20:32.622281 LibreELEC connmand[2411]: eth0 {newlink} index 3 operstate 6 <UP>

Feb 07 18:20:32.696291 LibreELEC connmand[2411]: ../src/dhcp.c:__connman_dhcp_start()

Feb 07 18:20:32.699878 LibreELEC connmand[2411]: DHCP index 3: switch listening mode (0 ==> 1)
Feb 07 18:20:32.700515 LibreELEC connmand[2411]: DHCP index 3: DHCP client start with state init_reboot
Feb 07 18:20:32.700587 LibreELEC connmand[2411]: DHCP index 3: sending DHCP request (state 1)

Feb 07 18:20:32.728618 LibreELEC connmand[2411]: ../src/ipconfig.c:disable_ipv6()

Feb 07 18:20:32.729959 LibreELEC connmand[2411]: eth0 {del} route fe80:: gw :: scope 0 <UNIVERSE>

Feb 07 18:20:38.334745 LibreELEC connmand[2411]: DHCP index 3: sending DHCP discover request

Feb 07 18:20:40.217224 LibreELEC systemd[1]: kodi-waitonnetwork.service: Failed with result 'exit-code'.

Feb 07 18:20:41.341668 LibreELEC connmand[2411]: DHCP index 3: received DHCP packet xid 0x17fecb4 (current state 0)
Feb 07 18:20:41.341824 LibreELEC connmand[2411]: DHCP index 3: init ip 10.42.0.133 secs 0 -> broadcast flag off
Feb 07 18:20:41.341860 LibreELEC connmand[2411]: DHCP index 3: start request (retries 0)
Feb 07 18:20:41.341890 LibreELEC connmand[2411]: DHCP index 3: sending DHCP request (state 2)
Feb 07 18:20:41.372578 LibreELEC connmand[2411]: DHCP index 3: received DHCP packet xid 0x17fecb4 (current state 2)
Feb 07 18:20:41.372806 LibreELEC connmand[2411]: DHCP index 3: switch listening mode (1 ==> 0)
```
After sending the first DHCP request at 18:20:32.700587 IPv6 is disabled at 18:20:32.728618. I'm 
wondering if the request is seen at the DHCP server and replied?

**Comment by mglae** [comment]
Add kernel parameter `ipv6.disable=1` to test if the IPv6 route does interfere.

**Comment by jernejsk** [comment]
That parameter unfortunately doesn't help either.

Here is log with that parameter http://ix.io/2OZm and wireshark dump, from boot onwards http://jernej.libreelec.tv/dump/dhcp.pcapng

It's interesting that it takes 5.7 seconds to send 2 dhcp packets from board. PC (DHCP server) needs 2 seconds to confirm requested address, which sounds about right.

**Comment by mglae** [comment]
The DHCP server is not responding to the DHCP request. After the timeout of 5 sec a DHCP discover is send. 

Which DHCP server are you using? Is there anything logged?

From the implementation the first DHCP request is from a special [REBOOTING state](https://git.kernel.org/pub/scm/network/connman/connman.git/commit/gdhcp/client.c?h=1.39&id=bd37ce0f629e1f69712710b57cf9d34833ab667f) used to speed up connection.

**Comment by jernejsk** [comment]
I configured shared connection in NetworkManager which in turn uses dnsmasq 2.84 for dhcp server. Not sure where to check logs.

**Comment by mglae** [comment]
I'm using ISC DHCP Server - another difference.

Can you boot a Generic image to check if there is the same delay?

**Comment by jernejsk** [comment]
I did following experiment:
1. replace dnsmasq with isc dhcp
2. stop networkmanager
3. manually set up packet forwarding, sharing interface and dhcp server

This resulted in following log (LE board side) http://ix.io/2P50 and wireshark dump (PC side) http://jernej.libreelec.tv/dump/dhcp-isc.pcapng 
DHCP seems much more responsive now. So, it seems like that networkmanager waits on some packet (which is most likely discarded) to actually activate interface - starting dhcp server, configuring interface, etc. which takes time.

In short, I would say issue is how NetworkManager manages internet sharing. Thanks for all the help!

**Comment by mglae** [comment]
Good to know the reason for the behavior.

dnsmask need to be configured to support the init-reboot state:
```
       -K, --dhcp-authoritative
              Should be set when dnsmasq is definitely the only DHCP server on a network.   For
              DHCPv4, it changes the behaviour from strict RFC compliance so that DHCP requests
              on unknown leases from unknown hosts are not ignored. This allows  new  hosts  to
              get  a  lease  without  a tedious timeout under all circumstances. It also allows
              dnsmasq to rebuild its lease database without each client needing to reacquire  a
              lease, if the database is lost. For DHCPv6 it sets the priority in replies to 255
              (the maximum) instead of 0 (the minimum).
```

**Comment by jernejsk** [comment]
Thanks! This helps with networkmanager/dnsmasq combination. It's slower (6.5 vs. 3.5 seconds) than manual setup according to systemd-analyze, but at least there is no timeout. So, this work as intended, but I guess I will have to explain users over and over to either disable wait (some run it without connection set up) or optimize their setup. I imagine a lot of routers use dnsmasq...

**Comment by mglae** [comment]
> to either disable wait (some run it without connection set up)

IMO a very exceptional setup, usually users require network connection on startup.

> I imagine a lot of routers use dnsmasq...

Using dnsmasq without --dhcp-authoritative violates RFC 2131, this should likely be no problem. Just checked DD-WRT: as expected --dhcp-authoritative is enabled by default.
